techniques/systemSettings/misc/clockConfiguration/4.0/clockConfiguration.st

```diff
@@ -0,0 +1,462 @@
+#####################################################################################
+# Copyright 2011 Normation SAS
+# Copyright 2015 Janos Mattyasovszky
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+######################################################
+# Configure the ntp
+# If ntp is not installed, install it (linux)
+# Change the server configuration, and restart the daemon
+# if the configuration changed
+# Caution : a missing feature in cfe prevents from adding ,0x1 at the end of the domain name
+# so it's only safe to use with ips for now
+
+bundle agent check_clock_configuration
+{
+
+  vars:
+      "ntpServers_all" slist => {&CLOCK_NTPSERVERS: { splitstring("&it&", " *, *", 100)};separator=", "&};
+      "ntpServers" slist => unique("ntpServers_all");
+
+      "target_timezone" string => ifelse( strcmp("&CLOCK_TIMEZONE&", "custom"), "&CLOCK_TIMEZONE_CUSTOM&", "&CLOCK_TIMEZONE&" );
+
+      "clock_hwclock_rtc" string => "&CLOCK_HWSYNC_RTC&";
+
+    windows::
+      # For windows, we must extract both ip and domain name values
+      #"ipv4" slist => grep("(?:\\d{1,3}\.){3}\\d{1,3}","ntpServers");
+
+      #domain slist => not one of the above
+      # check that list aren't empty using reglist
+
+      &if(CLOCK_FQDNNTP)&
+      "keytemp" string => join("\,0x1 ", ntpServers);
+      "serverskey" string => "${keytemp}\,0x1";
+      &else&
+      "serverskey" string => join(" ", ntpServers);
+      &endif&
+
+      "target_windows_timezone" string => "${tz_variables.tz_win[${target_timezone}]}";
+      "current_windows_timezone" string => execresult("\"${sys.winsysdir}\tzutil.exe\" /g", "useshell");
+
+    linux::
+      "linux_timezone"   string => "${tz_variables.tz_linux[${target_timezone}]}";
+
+  classes:
+
+      "clock_timezone_nochange"  expression => strcmp("&CLOCK_TIMEZONE&","dontchange");
+      "clock_timezone_valid" expression => isvariable("tz_variables.tz_linux[${target_timezone}]");
```

Comment:
Yeah, I was thinking about that. My idea was that the whole windows/linux part was not coming from the user but is generated and provided by the system, so of one is present, so is the other. But you are right, just to have all edges covered I'll modify it it check it that way to catch all possible issues.